File: code/2025/04/10/1035. 不相交的线.cpp
class Solution {
public:
	int maxUncrossedLines(vector<int>& a, vector<int>& b) {
		int n = a.size(), m = b.size();
		std::vector<int> f(m);
		for (int i = 0; i < n; i++) {
			std::vector<int> nf(m);
			for (int j = 0; j < m; j++) {
				if (a[i] == b[j]) {
					if (j) {
						nf[j] = std::max(nf[j], f[j - 1]);
					}
					nf[j]++;
				} else {
					nf[j] = f[j];
					if (j) {
						nf[j] = std::max(nf[j], nf[j - 1]);
					}
				}
			}
			std::swap(f, nf);
		}
		return f.back();
	}
};

File: code/2025/04/10/1035. 不相交的线.py
class Solution:
    def maxUncrossedLines(self, a: List[int], b: List[int]) -> int:
        @cache
        def dfs(i, j):
            if i < 0 or j < 0:
                return 0
            if a[i] == b[j]:
                return dfs(i - 1, j - 1) + 1
            return max(dfs(i - 1, j), dfs(i, j - 1))
        return dfs(len(a) - 1, len(b) - 1)